File: httpdns_flutter_demo/packages/httpdns_plugin/android/src/main/kotlin/com/example/httpdns_plugin/HttpdnsPlugin.kt
package com.example.httpdns_plugin

import android.content.Context
import android.util.Log
import androidx.annotation.NonNull

import com.alibaba.sdk.android.httpdns.HttpDns
import com.alibaba.sdk.android.httpdns.HttpDnsService
import com.alibaba.sdk.android.httpdns.InitConfig
import com.alibaba.sdk.android.httpdns.RequestIpType
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result


class HttpdnsPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private var appContext: Context? = null

    // Cached service keyed by accountId to avoid re-creating
    private var service: HttpDnsService? = null
    private var accountId: String? = null
    private var secretKey: String? = null
    private var aesSecretKey: String? = null

    // Desired states collected before build()
    private var desiredPersistentCacheEnabled: Boolean? = null
    private var desiredDiscardExpiredAfterSeconds: Int? = null
    private var desiredReuseExpiredIPEnabled: Boolean? = null
    private var desiredLogEnabled: Boolean? = null
    private var desiredHttpsEnabled: Boolean? = null
    private var desiredPreResolveAfterNetworkChanged: Boolean? = null





    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        appContext = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "httpdns_plugin")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        // Log every incoming call with method name and raw arguments
        try {
            Log.i("HttpdnsPlugin", "invoke method=${call.method}, args=${call.arguments}")
        } catch (_: Throwable) {
            Log.i("HttpdnsPlugin", "invoke method=${call.method}, args=<unprintable>")
        }
        when (call.method) {
            // Dart: init(accountId, secretKey?, aesSecretKey?) — only save states here
            "initialize" -> {
                val args = call.arguments as? Map<*, *> ?: emptyMap<String, Any>()
                val ctx = appContext
                if (ctx == null) {
                    result.error("no_context", "Android context not attached", null)
                    return
                }
                val accountAny = args["accountId"]
                val account = when (accountAny) {
                    is Int -> accountAny.toString()
                    is Long -> accountAny.toString()
                    is String -> accountAny
                    else -> null
                }
                val secret = (args["secretKey"] as? String)?.takeIf { it.isNotBlank() }
                val aes = (args["aesSecretKey"] as? String)?.takeIf { it.isNotBlank() }

                if (account.isNullOrBlank()) {
                    Log.w("HttpdnsPlugin", "initialize missing accountId")
                    result.success(false)
                    return
                }
                // Save desired states only; actual build happens on 'build'
                accountId = account
                secretKey = secret
                aesSecretKey = aes
                Log.d("HttpdnsPlugin", "initialize saved state, account=$account")
                result.success(true)
            }



            // Dart: setLogEnabled(enabled) — save desired, still no-op at build-time on Android SDK
            "setLogEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") == true
                desiredLogEnabled = enabled
                Log.d("HttpdnsPlugin", "setLogEnabled desired=$enabled (no-op on Android)")
                result.success(null)
            }

            // Dart: setHttpsRequestEnabled(enabled)
            "setHttpsRequestEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") == true
                desiredHttpsEnabled = enabled
                Log.d("HttpdnsPlugin", "https request desired=$enabled")
                result.success(null)
            }

            // Dart: setPersistentCacheIPEnabled(enabled, discardExpiredAfterSeconds?) — save desired
            "setPersistentCacheIPEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") == true
                val discard = call.argument<Int>("discardExpiredAfterSeconds")
                desiredPersistentCacheEnabled = enabled
                desiredDiscardExpiredAfterSeconds = discard
                Log.d("HttpdnsPlugin", "persistent cache desired=$enabled discard=$discard")
                result.success(null)
            }

            // Dart: setReuseExpiredIPEnabled(enabled) — save desired
            "setReuseExpiredIPEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") == true
                desiredReuseExpiredIPEnabled = enabled
                Log.d("HttpdnsPlugin", "reuse expired ip desired=$enabled")
                result.success(null)
            }

            // Dart: setPreResolveAfterNetworkChanged(enabled) — save desired (applied at build via InitConfig)
            "setPreResolveAfterNetworkChanged" -> {
                val enabled = call.argument<Boolean>("enabled") == true
                desiredPreResolveAfterNetworkChanged = enabled
                Log.d("HttpdnsPlugin", "preResolveAfterNetworkChanged desired=$enabled")
                result.success(null)
            }

            // Dart: setPreResolveHosts(hosts, ipType)
            "setPreResolveHosts" -> {
                val hosts = call.argument<List<String>>("hosts") ?: emptyList()
                val ipTypeStr = call.argument<String>("ipType") ?: "auto"
                val type = when (ipTypeStr.lowercase()) {
                    "ipv4", "v4" -> RequestIpType.v4
                    "ipv6", "v6" -> RequestIpType.v6
                    "both", "64" -> RequestIpType.both
                    else -> RequestIpType.auto
                }
                try {
                    service?.setPreResolveHosts(hosts, type)
                    Log.d("HttpdnsPlugin", "preResolve set for ${hosts.size} hosts, type=$type")
                } catch (t: Throwable) {
                    Log.w("HttpdnsPlugin", "setPreResolveHosts failed: ${t.message}")
                }
                result.success(null)
            }

            // Dart: getSessionId
            "getSessionId" -> {
                val sid = try { service?.getSessionId() } catch (_: Throwable) { null }
                result.success(sid)
            }

            // Dart: cleanAllHostCache
            "cleanAllHostCache" -> {
                try {
                    // Best-effort: empty list to clear all
                    service?.cleanHostCache(ArrayList())
                } catch (t: Throwable) {
                    Log.w("HttpdnsPlugin", "cleanAllHostCache failed: ${t.message}")
                }
                result.success(null)
            }

            // Dart: build() — construct InitConfig and acquire service using desired states
            "build" -> {
                val ctx = appContext
                val account = accountId
                if (ctx == null || account.isNullOrBlank()) {
                    result.success(false)
                    return
                }
                try {
                    val builder = InitConfig.Builder()

                    // Optional builder params
                    try { builder.javaClass.getMethod("setContext", Context::class.java).invoke(builder, ctx) } catch (_: Throwable) {}
                    try {
                        if (!secretKey.isNullOrBlank()) {
                            builder.javaClass.getMethod("setSecretKey", String::class.java).invoke(builder, secretKey)
                        }
                    } catch (_: Throwable) {}
                    try {
                        if (!aesSecretKey.isNullOrBlank()) {
                            builder.javaClass.getMethod("setAesSecretKey", String::class.java).invoke(builder, aesSecretKey)
                        }
                    } catch (_: Throwable) {}
                    // Prefer HTTPS if requested
                    try {
                        desiredHttpsEnabled?.let { en ->
                            builder.javaClass.getMethod("setEnableHttps", Boolean::class.javaPrimitiveType).invoke(builder, en)
                        }
                    } catch (_: Throwable) {}
                    // Apply preResolve-after-network-changed
                    try {
                        desiredPreResolveAfterNetworkChanged?.let { en ->
                            builder.javaClass.getMethod("setPreResolveAfterNetworkChanged", Boolean::class.javaPrimitiveType).invoke(builder, en)
                        }
                    } catch (_: Throwable) {}

                    builder.buildFor(account)

                    service = if (!secretKey.isNullOrBlank()) {
                        HttpDns.getService(ctx, account, secretKey)
                    } else {
                        HttpDns.getService(ctx, account)
                    }
                    // Apply desired runtime flags on the service (try multiple API names to be compatible across SDK versions)
                    desiredHttpsEnabled?.let { enabled ->
                        try {
                            val m = service?.javaClass?.getMethod("setHTTPSRequestEnabled", Boolean::class.javaPrimitiveType)
                            m?.invoke(service, enabled)
                        } catch (_: Throwable) { }
                    }
                    desiredPersistentCacheEnabled?.let { enabled ->
                        try {
                            val m1 = service?.javaClass?.getMethod("setCachedIPEnabled", Boolean::class.javaPrimitiveType)
                            m1?.invoke(service, enabled)
                        } catch (_: Throwable) {
                            try {
                                val m2 = service?.javaClass?.getMethod("setPersistentCacheIPEnabled", Boolean::class.javaPrimitiveType)
                                m2?.invoke(service, enabled)
                            } catch (_: Throwable) { }
                        }
                    }
                    desiredReuseExpiredIPEnabled?.let { enabled ->
                        try {
                            val m1 = service?.javaClass?.getMethod("setExpiredIPEnabled", Boolean::class.javaPrimitiveType)
                            m1?.invoke(service, enabled)
                        } catch (_: Throwable) {
                            try {
                                val m2 = service?.javaClass?.getMethod("setReuseExpiredIPEnabled", Boolean::class.javaPrimitiveType)
                                m2?.invoke(service, enabled)
                            } catch (_: Throwable) { }
                        }
                    }
                    Log.d("HttpdnsPlugin", "build completed for account=$account")
                    result.success(true)
                } catch (t: Throwable) {
                    Log.e("HttpdnsPlugin", "build failed: ${t.message}")
                    result.success(false)
                }
            }

            // Dart: resolveHostSyncNonBlocking(hostname, ipType, sdnsParams?, cacheKey?)
            "resolveHostSyncNonBlocking" -> {
                val hostname = call.argument<String>("hostname")
                if (hostname.isNullOrBlank()) {
                    result.success(mapOf("ipv4" to emptyList<String>(), "ipv6" to emptyList<String>()))
                    return
                }
                val ipTypeStr = call.argument<String>("ipType") ?: "auto"
                val type = when (ipTypeStr.lowercase()) {
                    "ipv4", "v4" -> RequestIpType.v4
                    "ipv6", "v6" -> RequestIpType.v6
                    "both", "64" -> RequestIpType.both
                    else -> RequestIpType.auto
                }
                try {
                    val svc = service ?: run {
                        val ctx = appContext
                        val acc = accountId
                        if (ctx != null && !acc.isNullOrBlank()) HttpDns.getService(ctx, acc) else null
                    }
                    val r = svc?.getHttpDnsResultForHostSyncNonBlocking(hostname, type)
                    val v4 = r?.ips?.toList() ?: emptyList()
                    val v6 = r?.ipv6s?.toList() ?: emptyList()
                    // 记录解析结果，便于排查：包含 host、请求类型以及返回的 IPv4/IPv6 列表
                    Log.d(
                        "HttpdnsPlugin",
                        "resolve result host=" + hostname + ", type=" + type +
                            ", ipv4=" + v4.joinToString(prefix = "[", postfix = "]") +
                            ", ipv6=" + v6.joinToString(prefix = "[", postfix = "]")
                    )
                    result.success(mapOf("ipv4" to v4, "ipv6" to v6))
                } catch (t: Throwable) {
                    Log.e("HttpdnsPlugin", "resolveHostSyncNonBlocking failed: ${t.message}")
                    result.success(mapOf("ipv4" to emptyList<String>(), "ipv6" to emptyList<String>()))
                }
            }

            // Legacy methods removed: preResolve / clearCache handled at app layer if needed

            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        service = null
        appContext = null
    }
}

File: httpdns_flutter_demo/packages/aliyun_httpdns/ios/Classes/HttpdnsPlugin.swift
import Flutter
import UIKit
import AlicloudHTTPDNS

public class AliyunHttpDnsPlugin: NSObject, FlutterPlugin {
  private var channel: FlutterMethodChannel!

  // Desired states saved until build()
  private var desiredAccountId: Int?
  private var desiredSecretKey: String?
  private var desiredAesSecretKey: String?

  private var desiredPersistentCacheEnabled: Bool?
  private var desiredDiscardExpiredAfterSeconds: Int?
  private var desiredReuseExpiredIPEnabled: Bool?
  private var desiredLogEnabled: Bool?
  private var desiredHttpsEnabled: Bool?
  private var desiredPreResolveAfterNetworkChanged: Bool?
  private var desiredIPRankingMap: [String: NSNumber]?



  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "aliyun_httpdns", binaryMessenger: registrar.messenger())
    let instance = AliyunHttpDnsPlugin()
    instance.channel = channel
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    // Dart: init(accountId, secretKey?, aesSecretKey?) — only save desired state
    case "initialize":
      let options = call.arguments as? [String: Any] ?? [:]
      let accountIdAny = options["accountId"]
      let secretKey = options["secretKey"] as? String
      let aesSecretKey = options["aesSecretKey"] as? String

      guard let accountId = (accountIdAny as? Int) ?? Int((accountIdAny as? String) ?? "") else {
        NSLog("AliyunHttpDns: initialize missing accountId")
        result(false)
        return
      }
      desiredAccountId = accountId
      desiredSecretKey = secretKey
      desiredAesSecretKey = aesSecretKey
      NSLog("AliyunHttpDns: initialize saved accountId=\(accountId)")
      result(true)



    // Dart: setLogEnabled(enabled) — save desired
    case "setLogEnabled":
      let args = call.arguments as? [String: Any]
      let enabled = (args?["enabled"] as? Bool) ?? false
      desiredLogEnabled = enabled
      NSLog("AliyunHttpDns: log desired=\(enabled)")
      result(nil)

    case "setHttpsRequestEnabled":
      let args = call.arguments as? [String: Any]
      let enabled = (args?["enabled"] as? Bool) ?? false
      desiredHttpsEnabled = enabled
      NSLog("AliyunHttpDns: https request desired=\(enabled)")
      result(nil)

    // Dart: setPersistentCacheIPEnabled(enabled, discardExpiredAfterSeconds?) — save desired
    case "setPersistentCacheIPEnabled":
      let args = call.arguments as? [String: Any]
      let enabled = (args?["enabled"] as? Bool) ?? false
      let discard = args?["discardExpiredAfterSeconds"] as? Int
      desiredPersistentCacheEnabled = enabled
      desiredDiscardExpiredAfterSeconds = discard
      NSLog("AliyunHttpDns: persistent cache desired=\(enabled) discard=\(discard ?? -1)")
      result(nil)

    // Dart: setReuseExpiredIPEnabled(enabled) — save desired
    case "setReuseExpiredIPEnabled":
      let args = call.arguments as? [String: Any]
      let enabled = (args?["enabled"] as? Bool) ?? false
      desiredReuseExpiredIPEnabled = enabled
      NSLog("AliyunHttpDns: reuse expired ip desired=\(enabled)")
      result(nil)

    case "setPreResolveAfterNetworkChanged":
      let args = call.arguments as? [String: Any]
      let enabled = (args?["enabled"] as? Bool) ?? false
      desiredPreResolveAfterNetworkChanged = enabled
      NSLog("AliyunHttpDns: preResolveAfterNetworkChanged desired=\(enabled)")
      result(nil)

    case "setIPRankingList":
      let args = call.arguments as? [String: Any]
      let hostPortMap = args?["hostPortMap"] as? [String: NSNumber]
      desiredIPRankingMap = hostPortMap
      NSLog("AliyunHttpDns: IP ranking list desired, hosts=\(hostPortMap?.keys.joined(separator: ", ") ?? "")")
      result(nil)

    case "setPreResolveHosts":
      let args = call.arguments as? [String: Any]
      let hosts = (args?["hosts"] as? [String]) ?? []
      let ipTypeStr = (args?["ipType"] as? String) ?? "auto"
      switch ipTypeStr.lowercased() {
      case "ipv4", "v4":
        HttpDnsService.sharedInstance().setPreResolveHosts(hosts, queryIPType: AlicloudHttpDNS_IPType.init(0))
      case "ipv6", "v6":
        HttpDnsService.sharedInstance().setPreResolveHosts(hosts, queryIPType: AlicloudHttpDNS_IPType.init(1))
      case "both", "64":
        HttpDnsService.sharedInstance().setPreResolveHosts(hosts, queryIPType: AlicloudHttpDNS_IPType.init(2))
      default:
        HttpDnsService.sharedInstance().setPreResolveHosts(hosts)
      }
      result(nil)

    case "getSessionId":
      let sid = HttpDnsService.sharedInstance().getSessionId()
      result(sid)

    case "cleanAllHostCache":
      HttpDnsService.sharedInstance().cleanAllHostCache()
      result(nil)

    // Dart: build() — construct service and apply desired states
    case "build":
      guard let accountId = desiredAccountId else {
        result(false)
        return
      }
      // Initialize singleton
      if let secret = desiredSecretKey, !secret.isEmpty {
        if let aes = desiredAesSecretKey, !aes.isEmpty {
          _ = HttpDnsService(accountID: accountId, secretKey: secret, aesSecretKey: aes)
        } else {
          _ = HttpDnsService(accountID: accountId, secretKey: secret)
        }
      } else {
        _ = HttpDnsService(accountID: accountId) // deprecated but acceptable fallback
      }
      let svc = HttpDnsService.sharedInstance()
      // Apply desired runtime flags
      if let enable = desiredPersistentCacheEnabled {
        if let discard = desiredDiscardExpiredAfterSeconds, discard >= 0 {
          svc.setPersistentCacheIPEnabled(enable, discardRecordsHasExpiredFor: TimeInterval(discard))
        } else {
          svc.setPersistentCacheIPEnabled(enable)
        }
      }
      if let enable = desiredReuseExpiredIPEnabled {
        svc.setReuseExpiredIPEnabled(enable)
      }
      if let enable = desiredLogEnabled {
        svc.setLogEnabled(enable)
      }
      if let enable = desiredHttpsEnabled {
        svc.setHTTPSRequestEnabled(enable)
      }

      if let en = desiredPreResolveAfterNetworkChanged {
        svc.setPreResolveAfterNetworkChanged(en)
      }
      if let ipRankingMap = desiredIPRankingMap, !ipRankingMap.isEmpty {
        svc.setIPRankingDatasource(ipRankingMap)
      }
      NSLog("AliyunHttpDns: build completed accountId=\(accountId)")
      result(true)

    // Dart: resolveHostSyncNonBlocking(hostname, ipType, sdnsParams?, cacheKey?)
    case "resolveHostSyncNonBlocking":
      guard let args = call.arguments as? [String: Any], let host = args["hostname"] as? String else {
        result(["ipv4": [], "ipv6": []])
        return
      }
      let ipTypeStr = (args["ipType"] as? String) ?? "auto"
      let sdnsParams = args["sdnsParams"] as? [String: String]
      let cacheKey = args["cacheKey"] as? String
      let type: HttpdnsQueryIPType
      switch ipTypeStr.lowercased() {
      case "ipv4", "v4": type = .ipv4
      case "ipv6", "v6": type = .ipv6
      case "both", "64": type = .both
      default: type = .auto
      }
      let svc = HttpDnsService.sharedInstance()
      var v4: [String] = []
      var v6: [String] = []
      if let params = sdnsParams, let key = cacheKey, let r = svc.resolveHostSyncNonBlocking(host, by: type, withSdnsParams: params, sdnsCacheKey: key) {
        if r.hasIpv4Address() { v4 = r.ips }
        if r.hasIpv6Address() { v6 = r.ipv6s }
      } else if let r = svc.resolveHostSyncNonBlocking(host, by: type) {
        if r.hasIpv4Address() { v4 = r.ips }
        if r.hasIpv6Address() { v6 = r.ipv6s }
      }
      result(["ipv4": v4, "ipv6": v6])

    // Legacy methods removed: preResolve / clearCache

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}


